File: scripts/providers/codex_session_provider.py
"""Runtime-only Codex transport selection. It never reads or writes task-loop state."""
import asyncio
import inspect
import json
import os
import sys

sys.path.insert(0, os.path.dirname(__file__))

from codex_session_dispatch import dispatch as dispatch_cli
from codex_session_dispatch import normalize_thread_id


def prepared(reason, **extra):
    result = {"status": "PREPARED_ONLY", "submitted": False, "reason": reason}
    result.update(extra)
    return result


def unsupported(action, reason, **extra):
    result = {"status": "UNSUPPORTED", "supported": False, "action": action, "reason": reason}
    result.update(extra)
    return result


def pending_creation(reason, **extra):
    result = {
        "status": "PENDING_CREATION",
        "submitted": False,
        "bound": False,
        "vendor": "codex",
        "threadId": None,
        "reason": reason,
    }
    result.update(extra)
    return result


def creation_failed(reason, **extra):
    result = {
        "status": "CREATION_FAILED",
        "submitted": False,
        "bound": False,
        "vendor": "codex",
        "threadId": None,
        "reason": reason,
    }
    result.update(extra)
    return result


def normalize_request(request=None):
    request = request if isinstance(request, dict) else {}
    model_config = request.get("model_config") if isinstance(request.get("model_config"), dict) else {}
    thread = normalize_thread_id(request)
    return {
        "thread": thread,
        "threadId": thread,
        "message": request.get("message") or request.get("prompt"),
        "mode": "resume" if request.get("mode") == "resume" else "queue",
        "model": request.get("model") or model_config.get("model"),
        "reasoning_effort": request.get("reasoning_effort") or request.get("reasoningEffort") or request.get("thinking") or model_config.get("reasoning_effort"),
        "dry_run": bool(request.get("dry_run") or request.get("dryRun")),
    }


def normalize_create_request(request=None):
    request = request if isinstance(request, dict) else {}
    model_config = request.get("model_config") if isinstance(request.get("model_config"), dict) else {}
    return {
        "title": request.get("title") or request.get("name") or "",
        "prompt": request.get("prompt") or request.get("message") or "",
        "model": request.get("model") or model_config.get("model"),
        "reasoning_effort": request.get("reasoning_effort") or request.get("reasoningEffort") or request.get("thinking") or model_config.get("reasoning_effort"),
        "target": request.get("target"),
        "projectId": request.get("projectId") or request.get("project_id"),
        "environment": request.get("environment"),
        "role": request.get("role"),
        "dry_run": bool(request.get("dry_run") or request.get("dryRun")),
    }


def _collect_structured_objects(value, objects=None, seen=None):
    objects = objects if objects is not None else []
    seen = seen if seen is not None else set()
    if isinstance(value, str):
        try:
            parsed = json.loads(value)
            return objects if parsed == value else _collect_structured_objects(parsed, objects, seen)
        except (TypeError, ValueError):
            return objects
    if not isinstance(value, (dict, list)) or id(value) in seen:
        return objects
    seen.add(id(value))
    if isinstance(value, list):
        for item in value:
            _collect_structured_objects(item, objects, seen)
        return objects
    objects.append(value)
    for key in ("structuredContent", "response", "thread", "result", "data"):
        nested = value.get(key)
        if isinstance(nested, str):
            try:
                _collect_structured_objects(json.loads(nested), objects, seen)
            except (TypeError, ValueError):
                pass
        else:
            _collect_structured_objects(nested, objects, seen)
    content = value.get("content")
    if isinstance(content, list):
        for block in content:
            if not isinstance(block, dict) or block.get("type") != "text" or not isinstance(block.get("text"), str):
                continue
            try:
                _collect_structured_objects(json.loads(block["text"]), objects, seen)
            except (TypeError, ValueError):
                pass
    return objects


def _first_result_field(result, fields):
    for obj in _collect_structured_objects(result):
        for field in fields:
            value = obj.get(field)
            if value is not None and str(value).strip():
                return str(value).strip()
    return None


def formal_create_thread_id(result):
    return _first_result_field(result, ("threadId", "thread_id"))


def formal_client_thread_id(result):
    return _first_result_field(result, ("clientThreadId", "client_thread_id"))


def normalize_create_result(result, request, transport="host"):
    normalized_request = normalize_create_request(request)
    thread_id = formal_create_thread_id(result)
    if thread_id:
        return {
            "status": "READY",
            "submitted": True,
            "bound": False,
            "vendor": "codex",
            "threadId": thread_id,
            "id_kind": "threadId",
            "transport": transport,
            "request": normalized_request,
            "result": result,
        }
    client_thread_id = formal_client_thread_id(result)
    if client_thread_id:
        return pending_creation(
            "Codex host returned clientThreadId; wait for the formal threadId before binding or sending",
            transport=transport,
            client_thread_id=client_thread_id,
            request=normalized_request,
            host_result=result,
        )
    return pending_creation(
        "Codex creation was requested but the host has not returned a formal threadId",
        transport=transport,
        request=normalized_request,
        host_result=result,
    )


def _create_adapter_entries(capabilities):
    capabilities = capabilities if isinstance(capabilities, dict) else {}
    entries = []
    create_adapter = capabilities.get("create")
    if callable(create_adapter):
        entries.append((capabilities.get("transport") or "host", create_adapter))
    for transport in ("desktop", "sdk", "api"):
        direct = capabilities.get(f"{transport}Create")
        nested = capabilities.get(transport)
        nested = nested.get("create") if isinstance(nested, dict) else None
        if callable(direct):
            entries.append((transport, direct))
        elif callable(nested):
            entries.append((transport, nested))
    return entries


def create(request, capabilities=None, **options):
    normalized = normalize_create_request(request)
    if options.get("dry_run") or options.get("dryRun"):
        normalized["dry_run"] = True
    if not normalized["title"] or not normalized["prompt"]:
        return creation_failed("title and prompt are required")
    if normalized["dry_run"]:
        return pending_creation("dry-run; no Codex host transport was executed", request=normalized)

    entries = _create_adapter_entries(capabilities)
    if not entries:
        return pending_creation(
            "Codex Desktop creation must be completed by the host; this repository cannot invent or persist a threadId",
            request=normalized,
            creation_request=normalized,
            host_action="create_thread",
        )

    failures = []
    for transport, adapter in entries:
        try:
            result = adapter({**normalized, "transport": transport})
            if inspect.isawaitable(result):
                failures.append(f"Codex {transport} adapter requires create_async()")
                continue
            outcome = normalize_create_result(result, normalized, transport)
            if outcome.get("status") in ("READY", "PENDING_CREATION"):
                return outcome
            failures.append(f"{transport}: {outcome.get('reason', 'adapter rejected creation')}")
        except Exception as error:
            failures.append(f"Codex {transport} adapter failed: {error}")
    return creation_failed("; ".join(failures) or "No Codex creation adapter returned a result", request=normalized)


async def create_async(request, capabilities=None, **options):
    normalized = normalize_create_request(request)
    if options.get("dry_run") or options.get("dryRun"):
        normalized["dry_run"] = True
    if not normalized["title"] or not normalized["prompt"]:
        return creation_failed("title and prompt are required")
    if normalized["dry_run"]:
        return pending_creation("dry-run; no Codex host transport was executed", request=normalized)

    entries = _create_adapter_entries(capabilities)
    if not entries:
        return pending_creation(
            "Codex Desktop creation must be completed by the host; this repository cannot invent or persist a threadId",
            request=normalized,
            creation_request=normalized,
            host_action="create_thread",
        )

    failures = []
    for transport, adapter in entries:
        try:
            result = adapter({**normalized, "transport": transport})
            if inspect.isawaitable(result):
                result = await result
            outcome = normalize_create_result(result, normalized, transport)
            if outcome.get("status") in ("READY", "PENDING_CREATION"):
                return outcome
            failures.append(f"{transport}: {outcome.get('reason', 'adapter rejected creation')}")
        except Exception as error:
            failures.append(f"Codex {transport} adapter failed: {error}")
    return creation_failed("; ".join(failures) or "No Codex creation adapter returned a result", request=normalized)


def _lifecycle_adapter_entries(action, capabilities):
    capabilities = capabilities if isinstance(capabilities, dict) else {}
    entries = []
    direct = capabilities.get(action)
    if callable(direct):
        entries.append((capabilities.get("transport") or "host", direct))
    suffix = action[0].upper() + action[1:]
    for transport in ("desktop", "sdk", "api"):
        direct = capabilities.get(f"{transport}{suffix}")
        nested = capabilities.get(transport)
        nested = nested.get(action) if isinstance(nested, dict) else None
        if callable(direct):
            entries.append((transport, direct))
        elif callable(nested):
            entries.append((transport, nested))
    return entries


def normalize_lifecycle_result(action, result, transport, thread_id):
    if isinstance(result, dict) and result.get("status"):
        normalized = dict(result)
        normalized.update({"action": action, "transport": transport, "threadId": thread_id})
        return normalized
    return {"status": "READY", "action": action, "transport": transport, "threadId": thread_id, "result": result}


def read(request, capabilities=None):
    thread_id = normalize_thread_id(request)
    if not thread_id:
        return unsupported("read", "formal Codex threadId is required; clientThreadId cannot be read")
    entries = _lifecycle_adapter_entries("read", capabilities)
    if not entries:
        return unsupported("read", "No Codex read adapter is connected", threadId=thread_id)
    transport, adapter = entries[0]
    try:
        result = adapter({"threadId": thread_id, "thread": thread_id, "transport": transport})
        if inspect.isawaitable(result):
            return prepared("Codex read adapter requires read_async()", action="read", threadId=thread_id)
        return normalize_lifecycle_result("read", result, transport, thread_id)
    except Exception as error:
        return {"status": "READ_FAILED", "action": "read", "threadId": thread_id, "reason": str(error)}


async def read_async(request, capabilities=None):
    thread_id = normalize_thread_id(request)
    if not thread_id:
        return unsupported("read", "formal Codex threadId is required; clientThreadId cannot be read")
    entries = _lifecycle_adapter_entries("read", capabilities)
    if not entries:
        return unsupported("read", "No Codex read adapter is connected", threadId=thread_id)
    transport, adapter = entries[0]
    try:
        result = adapter({"threadId": thread_id, "thread": thread_id, "transport": transport})
        if inspect.isawaitable(result):
            result = await result
        return normalize_lifecycle_result("read", result, transport, thread_id)
    except Exception as error:
        return {"status": "READ_FAILED", "action": "read", "threadId": thread_id, "reason": str(error)}


async def wait(request, capabilities=None):
    thread_id = normalize_thread_id(request)
    if not thread_id:
        return unsupported("wait", "formal Codex threadId is required; clientThreadId cannot be waited on")
    entries = _lifecycle_adapter_entries("wait", capabilities)
    if not entries:
        return unsupported("wait", "No Codex wait adapter is connected", threadId=thread_id)
    transport, adapter = entries[0]
    try:
        result = adapter({"threadId": thread_id, "thread": thread_id, "transport": transport})
        if inspect.isawaitable(result):
            result = await result
        return normalize_lifecycle_result("wait", result, transport, thread_id)
    except Exception as error:
        return {"status": "WAIT_FAILED", "action": "wait", "threadId": thread_id, "reason": str(error)}


def _send_adapter_entries(capabilities):
    capabilities = capabilities if isinstance(capabilities, dict) else {}
    entries = []
    send_adapter = capabilities.get("send")
    if callable(send_adapter):
        entries.append((capabilities.get("transport") or "host", send_adapter))
    for transport in ("desktop", "sdk", "api"):
        direct = capabilities.get(transport)
        named = capabilities.get(f"{transport}Send")
        nested = direct.get("send") if isinstance(direct, dict) else None
        if callable(direct):
            entries.append((transport, direct))
        elif callable(named):
            entries.append((transport, named))
        elif callable(nested):
            entries.append((transport, nested))
    return entries


def get_capabilities(capabilities=None):
    has_action = lambda action: bool(_lifecycle_adapter_entries(action, capabilities))
    has_create = bool(_create_adapter_entries(capabilities))
    normalized_capabilities = capabilities if isinstance(capabilities, dict) else {}
    has_send = bool(_send_adapter_entries(normalized_capabilities)) or normalized_capabilities.get("cli") is True or callable(normalized_capabilities.get("run_cli"))
    return {
        "vendor": "codex",
        "create": {"supported": has_create, "status": "AVAILABLE" if has_create else "PENDING_CREATION"},
        "send": {"supported": has_send, "status": "AVAILABLE" if has_send else "UNSUPPORTED"},
        "read": {"supported": has_action("read"), "status": "AVAILABLE" if has_action("read") else "UNSUPPORTED"},
        "wait": {"supported": has_action("wait"), "status": "AVAILABLE" if has_action("wait") else "UNSUPPORTED"},
        "formal_thread_id_required": True,
        "client_thread_id_routable": False,
        "host_managed_creation": not has_create,
    }


def confirmed(transport, result):
    if isinstance(result, dict) and result.get("submitted") is True:
        return {"status": "SUBMITTED", "submitted": True, "transport": transport, "result": result}
    return prepared("Codex adapter did not return explicit submission confirmation", transport=transport, result=result)


def submit(request, capabilities=None, cli=True, run_cli=None):
    normalized = normalize_request(request)
    if not normalized["thread"] or not normalized["message"]:
        return prepared("thread and message are required")
    if normalized["dry_run"]:
        return prepared("dry-run; no Codex transport was executed", request=normalized)
    capabilities = capabilities or {}
    failures = []
    for transport, adapter in _send_adapter_entries(capabilities):
        if inspect.iscoroutinefunction(adapter):
            failures.append(f"Codex {transport} adapter requires submit_async()")
            continue
        try:
            result = adapter({**normalized, "transport": transport})
            if inspect.isawaitable(result):
                failures.append(f"Codex {transport} adapter returned an awaitable; use submit_async()")
                continue
            outcome = confirmed(transport, result)
            if outcome.get("submitted") is True:
                return outcome
            failures.append(f"{transport}: {outcome.get('reason', 'adapter rejected submission')}")
        except Exception as error:
            failures.append(f"Codex {transport} adapter failed: {error}")
    if cli:
        return dispatch_cli(normalized, run_cli or capabilities.get("run_cli"))
    return prepared("; ".join(failures) or "No Codex Desktop, SDK, API, or CLI transport is available")


async def submit_async(request, capabilities=None, cli=True, run_cli=None):
    capabilities = capabilities or {}
    normalized = normalize_request(request)
    if not normalized["thread"] or not normalized["message"]:
        return prepared("thread and message are required")
    if normalized["dry_run"]:
        return prepared("dry-run; no Codex transport was executed", request=normalized)
    for transport, adapter in _send_adapter_entries(capabilities):
        try:
            result = adapter({**normalized, "transport": transport})
            if inspect.isawaitable(result):
                result = await result
            outcome = confirmed(transport, result)
            if outcome.get("submitted") is True:
                return outcome
        except Exception as error:
            continue
    return submit(normalized, {}, cli, run_cli or capabilities.get("run_cli"))


send = submit

File: scripts/providers/codex_session_provider.js
'use strict';

const cliDispatch = require('./codex_session_dispatch');

function prepared(reason, extra = {}) {
  return { status: 'PREPARED_ONLY', submitted: false, reason, ...extra };
}

function unsupported(action, reason, extra = {}) {
  return { status: 'UNSUPPORTED', supported: false, action, reason, ...extra };
}

function pendingCreation(reason, extra = {}) {
  return {
    status: 'PENDING_CREATION',
    submitted: false,
    bound: false,
    vendor: 'codex',
    threadId: null,
    reason,
    ...extra
  };
}

function creationFailed(reason, extra = {}) {
  return {
    status: 'CREATION_FAILED',
    submitted: false,
    bound: false,
    vendor: 'codex',
    threadId: null,
    reason,
    ...extra
  };
}

function normalizeRequest(request = {}) {
  request = request && typeof request === 'object' && !Array.isArray(request) ? request : {};
  const modelConfig = request.model_config && typeof request.model_config === 'object' ? request.model_config : {};
  const thread = cliDispatch.normalizeThreadId(request);
  return {
    thread,
    threadId: thread,
    message: request.message || request.prompt,
    mode: request.mode === 'resume' ? 'resume' : 'queue',
    model: request.model || modelConfig.model,
    reasoning_effort: request.reasoning_effort || request.reasoningEffort || request.thinking || modelConfig.reasoning_effort,
    dryRun: Boolean(request.dryRun || request.dry_run)
  };
}

function normalizeCreateRequest(request = {}) {
  request = request && typeof request === 'object' && !Array.isArray(request) ? request : {};
  const modelConfig = request.model_config && typeof request.model_config === 'object' ? request.model_config : {};
  return {
    title: request.title || request.name || '',
    prompt: request.prompt || request.message || '',
    model: request.model || modelConfig.model,
    reasoning_effort: request.reasoning_effort || request.reasoningEffort || request.thinking || modelConfig.reasoning_effort,
    target: request.target,
    projectId: request.projectId || request.project_id,
    environment: request.environment,
    role: request.role,
    dryRun: Boolean(request.dryRun || request.dry_run)
  };
}

function collectStructuredObjects(value, objects = [], seen = new Set()) {
  if (typeof value === 'string') {
    try {
      const parsed = JSON.parse(value);
      return parsed === value ? objects : collectStructuredObjects(parsed, objects, seen);
    } catch (error) {
      return objects;
    }
  }
  if (value == null || typeof value !== 'object' || seen.has(value)) return objects;
  seen.add(value);
  if (Array.isArray(value)) {
    value.forEach(item => collectStructuredObjects(item, objects, seen));
    return objects;
  }
  objects.push(value);
  for (const key of ['structuredContent', 'response', 'thread', 'result', 'data']) {
    const nested = value[key];
    if (typeof nested === 'string') {
      try {
        collectStructuredObjects(JSON.parse(nested), objects, seen);
      } catch (error) {
        // Non-JSON text is not a structured Codex result.
      }
    } else {
      collectStructuredObjects(nested, objects, seen);
    }
  }
  if (Array.isArray(value.content)) {
    for (const block of value.content) {
      if (!block || block.type !== 'text' || typeof block.text !== 'string') continue;
      try {
        collectStructuredObjects(JSON.parse(block.text), objects, seen);
      } catch (error) {
        // Content blocks may contain ordinary human-readable text.
      }
    }
  }
  return objects;
}

function firstResultField(result, fields) {
  for (const object of collectStructuredObjects(result)) {
    for (const field of fields) {
      const value = object[field];
      if (value != null && String(value).trim()) return String(value).trim();
    }
  }
  return null;
}

function formalCreateThreadId(result) {
  return firstResultField(result, ['threadId', 'thread_id']);
}

function formalClientThreadId(result) {
  return firstResultField(result, ['clientThreadId', 'client_thread_id']);
}

function normalizeCreateResult(result, request, transport = 'host') {
  const normalizedRequest = normalizeCreateRequest(request);
  const threadId = formalCreateThreadId(result);
  if (threadId) {
    return {
      status: 'READY',
      submitted: true,
      bound: false,
      vendor: 'codex',
      threadId,
      id_kind: 'threadId',
      transport,
      request: normalizedRequest,
      result
    };
  }
  const clientThreadId = formalClientThreadId(result);
  if (clientThreadId) {
    return pendingCreation(
      'Codex host returned clientThreadId; wait for the formal threadId before binding or sending',
      {
        transport,
        client_thread_id: clientThreadId,
        request: normalizedRequest,
        host_result: result
      }
    );
  }
  return pendingCreation(
    'Codex creation was requested but the host has not returned a formal threadId',
    { transport, request: normalizedRequest, host_result: result }
  );
}

function createAdapterEntries(capabilities = {}) {
  const entries = [];
  if (typeof capabilities.create === 'function') {
    entries.push({ transport: capabilities.transport || 'host', adapter: capabilities.create });
  }
  for (const transport of ['desktop', 'sdk', 'api']) {
    const direct = capabilities[`${transport}Create`];
    const nested = capabilities[transport] && typeof capabilities[transport] === 'object'
      ? capabilities[transport].create
      : null;
    if (typeof direct === 'function') entries.push({ transport, adapter: direct });
    else if (typeof nested === 'function') entries.push({ transport, adapter: nested });
  }
  return entries;
}

function create(request, capabilities = {}, options = {}) {
  const normalized = normalizeCreateRequest(request);
  if (options.dryRun) normalized.dryRun = true;
  if (!normalized.title || !normalized.prompt) return creationFailed('title and prompt are required');
  if (normalized.dryRun) return pendingCreation('dry-run; no Codex host transport was executed', { request: normalized });

  const entries = createAdapterEntries(capabilities);
  if (!entries.length) {
    return pendingCreation(
      'Codex Desktop creation must be completed by the host; this repository cannot invent or persist a threadId',
      { request: normalized, creation_request: normalized, host_action: 'create_thread' }
    );
  }

  const failures = [];
  for (const entry of entries) {
    try {
      const result = entry.adapter({ ...normalized, transport: entry.transport });
      if (result && typeof result.then === 'function') {
        failures.push(`Codex ${entry.transport} adapter requires createAsync()`);
        continue;
      }
      const outcome = normalizeCreateResult(result, normalized, entry.transport);
      if (outcome.status === 'READY' || outcome.status === 'PENDING_CREATION') return outcome;
      failures.push(`${entry.transport}: ${outcome.reason}`);
    } catch (error) {
      failures.push(`Codex ${entry.transport} adapter failed: ${error.message}`);
    }
  }
  return creationFailed(failures.join('; ') || 'No Codex creation adapter returned a result', { request: normalized });
}

async function createAsync(request, capabilities = {}, options = {}) {
  const normalized = normalizeCreateRequest(request);
  if (options.dryRun) normalized.dryRun = true;
  if (!normalized.title || !normalized.prompt) return creationFailed('title and prompt are required');
  if (normalized.dryRun) return pendingCreation('dry-run; no Codex host transport was executed', { request: normalized });

  const entries = createAdapterEntries(capabilities);
  if (!entries.length) {
    return pendingCreation(
      'Codex Desktop creation must be completed by the host; this repository cannot invent or persist a threadId',
      { request: normalized, creation_request: normalized, host_action: 'create_thread' }
    );
  }

  const failures = [];
  for (const entry of entries) {
    try {
      const result = await entry.adapter({ ...normalized, transport: entry.transport });
      const outcome = normalizeCreateResult(result, normalized, entry.transport);
      if (outcome.status === 'READY' || outcome.status === 'PENDING_CREATION') return outcome;
      failures.push(`${entry.transport}: ${outcome.reason}`);
    } catch (error) {
      failures.push(`Codex ${entry.transport} adapter failed: ${error.message}`);
    }
  }
  return creationFailed(failures.join('; ') || 'No Codex creation adapter returned a result', { request: normalized });
}

function lifecycleAdapterEntries(action, capabilities = {}) {
  const entries = [];
  if (typeof capabilities[action] === 'function') entries.push({ transport: capabilities.transport || 'host', adapter: capabilities[action] });
  for (const transport of ['desktop', 'sdk', 'api']) {
    const direct = capabilities[`${transport}${action[0].toUpperCase()}${action.slice(1)}`];
    const nested = capabilities[transport] && typeof capabilities[transport] === 'object'
      ? capabilities[transport][action]
      : null;
    if (typeof direct === 'function') entries.push({ transport, adapter: direct });
    else if (typeof nested === 'function') entries.push({ transport, adapter: nested });
  }
  return entries;
}

function normalizeLifecycleResult(action, result, transport, threadId) {
  if (result && typeof result === 'object' && result.status) return { ...result, action, transport, threadId };
  return { status: 'READY', action, transport, threadId, result };
}

function read(request, capabilities = {}) {
  const threadId = cliDispatch.normalizeThreadId(request);
  if (!threadId) return unsupported('read', 'formal Codex threadId is required; clientThreadId cannot be read');
  const entries = lifecycleAdapterEntries('read', capabilities);
  if (!entries.length) return unsupported('read', 'No Codex read adapter is connected', { threadId });
  try {
    const result = entries[0].adapter({ threadId, thread: threadId, transport: entries[0].transport });
    if (result && typeof result.then === 'function') return prepared('Codex read adapter requires readAsync()', { action: 'read', threadId });
    return normalizeLifecycleResult('read', result, entries[0].transport, threadId);
  } catch (error) {
    return { status: 'READ_FAILED', action: 'read', threadId, reason: error.message };
  }
}

async function readAsync(request, capabilities = {}) {
  const threadId = cliDispatch.normalizeThreadId(request);
  if (!threadId) return unsupported('read', 'formal Codex threadId is required; clientThreadId cannot be read');
  const entries = lifecycleAdapterEntries('read', capabilities);
  if (!entries.length) return unsupported('read', 'No Codex read adapter is connected', { threadId });
  try {
    const entry = entries[0];
    const result = await entry.adapter({ threadId, thread: threadId, transport: entry.transport });
    return normalizeLifecycleResult('read', result, entry.transport, threadId);
  } catch (error) {
    return { status: 'READ_FAILED', action: 'read', threadId, reason: error.message };
  }
}

async function wait(request, capabilities = {}) {
  const threadId = cliDispatch.normalizeThreadId(request);
  if (!threadId) return unsupported('wait', 'formal Codex threadId is required; clientThreadId cannot be waited on');
  const entries = lifecycleAdapterEntries('wait', capabilities);
  if (!entries.length) return unsupported('wait', 'No Codex wait adapter is connected', { threadId });
  try {
    const result = await entries[0].adapter({ threadId, thread: threadId, transport: entries[0].transport });
    return normalizeLifecycleResult('wait', result, entries[0].transport, threadId);
  } catch (error) {
    return { status: 'WAIT_FAILED', action: 'wait', threadId, reason: error.message };
  }
}

function sendAdapterEntries(capabilities = {}) {
  const entries = [];
  if (typeof capabilities.send === 'function') {
    entries.push({ transport: capabilities.transport || 'host', adapter: capabilities.send });
  }
  for (const transport of ['desktop', 'sdk', 'api']) {
    const direct = capabilities[transport];
    const named = capabilities[`${transport}Send`];
    const nested = direct && typeof direct === 'object' ? direct.send : null;
    if (typeof direct === 'function') entries.push({ transport, adapter: direct });
    else if (typeof named === 'function') entries.push({ transport, adapter: named });
    else if (typeof nested === 'function') entries.push({ transport, adapter: nested });
  }
  return entries;
}

function getCapabilities(capabilities = {}) {
  const hasAction = action => lifecycleAdapterEntries(action, capabilities).length > 0;
  const hasCreate = createAdapterEntries(capabilities).length > 0;
  const hasSend = sendAdapterEntries(capabilities).length > 0
    || capabilities.cli === true
    || typeof capabilities.runCli === 'function';
  return {
    vendor: 'codex',
    create: { supported: hasCreate, status: hasCreate ? 'AVAILABLE' : 'PENDING_CREATION' },
    send: { supported: hasSend, status: hasSend ? 'AVAILABLE' : 'UNSUPPORTED' },
    read: { supported: hasAction('read'), status: hasAction('read') ? 'AVAILABLE' : 'UNSUPPORTED' },
    wait: { supported: hasAction('wait'), status: hasAction('wait') ? 'AVAILABLE' : 'UNSUPPORTED' },
    formal_thread_id_required: true,
    client_thread_id_routable: false,
    host_managed_creation: !hasCreate
  };
}

function confirmed(transport, result) {
  return result && result.submitted === true
    ? { status: 'SUBMITTED', submitted: true, transport, result }
    : prepared('Codex adapter did not return explicit submission confirmation', { transport, result });
}

function submit(request, capabilities = {}, options = {}) {
  const normalized = normalizeRequest(request);
  if (!normalized.thread || !normalized.message) return prepared('thread and message are required');
  if (normalized.dryRun) return prepared('dry-run; no Codex transport was executed', { request: normalized });

  const failures = [];
  for (const entry of sendAdapterEntries(capabilities)) {
    const { transport, adapter } = entry;
    if (adapter.constructor && adapter.constructor.name === 'AsyncFunction') {
      failures.push(`Codex ${transport} adapter requires submitAsync()`);
      continue;
    }
    try {
      const result = adapter({ ...normalized, transport });
      if (result && typeof result.then === 'function') {
        failures.push(`Codex ${transport} adapter returned a thenable; use submitAsync()`);
        continue;
      }
      const outcome = confirmed(transport, result);
      if (outcome.submitted === true) return outcome;
      failures.push(`${transport}: ${outcome.reason}`);
    } catch (error) {
      failures.push(`Codex ${transport} adapter failed: ${error.message}`);
    }
  }
  return options.cli !== false
    ? cliDispatch.dispatch(normalized, options.runCli || capabilities.runCli)
    : prepared(failures.join('; ') || 'No Codex Desktop, SDK, API, or CLI transport is available');
}

async function submitAsync(request, capabilities = {}, options = {}) {
  capabilities = capabilities || {};
  const normalized = normalizeRequest(request);
  if (!normalized.thread || !normalized.message) return prepared('thread and message are required');
  if (normalized.dryRun) return prepared('dry-run; no Codex transport was executed', { request: normalized });

  const failures = [];
  for (const entry of sendAdapterEntries(capabilities)) {
    const { transport, adapter } = entry;
    try {
      const outcome = confirmed(transport, await adapter({ ...normalized, transport }));
      if (outcome.submitted === true) return outcome;
      failures.push(`${transport}: ${outcome.reason}`);
    } catch (error) {
      failures.push(`Codex ${transport} adapter failed: ${error.message}`);
    }
  }
  const fallbackOptions = { ...options, runCli: options.runCli || capabilities.runCli };
  return submit(normalized, {}, fallbackOptions);
}

module.exports = {
  normalizeRequest,
  normalizeCreateRequest,
  formalCreateThreadId,
  formalClientThreadId,
  normalizeCreateResult,
  normalizeThreadId: cliDispatch.normalizeThreadId,
  create,
  createAsync,
  submit,
  send: submit,
  submitAsync,
  read,
  readAsync,
  wait,
  getCapabilities,
  prepared,
  pendingCreation,
  creationFailed,
  unsupported
};
